lib/pbench/server/filetree.py

```diff
@@ -299,18 +313,35 @@ def unpack(self, incoming: Path, results: Path):
         ensured by calling this indirectly through the Controller class unpack
         method.
 
-        TODO: This is a prototype for testing, and doesn't actually unpack
-        the tarball as we're going to be relying on the current unpack pipeline
-        command for some time.
-
         Args:
             incoming: Controller's directory in the INCOMING tree
             results: Controller's directory in the RESULTS tree
+
+        Raises:
+            TarballUnpackError if an exception occurs or a problem is
+            detected while unpacking the tarball.
         """
-        self.logger.warning("Tarball unpack is not yet implemented")
         unpacked = incoming / self.name
-        unpacked.mkdir(parents=True)  # Just create an empty directory for now
+        unpacked.mkdir(parents=True)
         self.unpacked = unpacked
+
+        try:
+            process = subprocess.run(
+                ["tar", "-xf", self.tarball_path],
+                cwd=self.unpacked,
+                stdin=subprocess.DEVNULL,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+                universal_newlines=True,
+            )
+        except Exception as exc:
+            raise TarballUnpackError(self.tarball_path.name, exc) from exc
+        else:
+            if process.returncode != 0:
+                raise TarballUnpackError(
+                    self.tarball_path.name,
+                    f"{process.stderr}, returncode: {process.returncode}",
```

Comment:
While I like reporting the `stderr` here, this `raise` doesn't conform to the exception's constructor signature since the second parameter is supposed to be an `Exception`.

The constructor is just taking the `str()` of the parameter anyway, so maybe just change it to specify `message: str`; that'll match this use, and for the other you'd pass `str(exc)`.

You might also consider making this a bit more clear by using something like `f"stderr {process.stderr!r}, returncode {process.returncode}"`